pkg/kubelet/kubelet_node_status.go

```diff
@@ -554,3 +568,56 @@ func validateNodeIP(nodeIP net.IP) error {
 	}
 	return fmt.Errorf("Node IP: %q not found in the host's network interfaces", nodeIP.String())
 }
+
+// nodeStatusHasChanged compares the original node and current node's status and
+// returns true if any change happens. The heartbeat timestamp is ignored.
+func nodeStatusHasChanged(originalStatus *v1.NodeStatus, status *v1.NodeStatus) bool {
+	if originalStatus == nil || status == nil {
+		return !apiequality.Semantic.DeepEqual(originalStatus, status)
+	}
+
+	originalStatusElem := reflect.ValueOf(originalStatus).Elem()
+	statusElem := reflect.ValueOf(status).Elem()
+
+	for i := 0; i < statusElem.NumField(); i++ {
+		originalValue := originalStatusElem.Field(i).Interface()
+		value := statusElem.Field(i).Interface()
+		if statusElem.Type().Field(i).Name == "Conditions" {
+			originalConditions := originalValue.([]v1.NodeCondition)
+			conditions := value.([]v1.NodeCondition)
+			if nodeConditionsHaveChanged(originalConditions, conditions) {
+				return true
+			}
+		} else if !apiequality.Semantic.DeepEqual(originalValue, value) {
+			return true
+		}
+	}
+	return false
+}
+
+// nodeConditionsHaveChanged compares the original node and current node's
+// conditions and returns true if any change happens. The heartbeat timestamp is
+// ignored.
+func nodeConditionsHaveChanged(originalConditions []v1.NodeCondition, conditions []v1.NodeCondition) bool {
+	if len(originalConditions) != len(conditions) {
+		return true
+	}
+
+	originalConditionsCopy := make([]v1.NodeCondition, len(originalConditions))
+	copy(originalConditionsCopy, originalConditions)
+	conditionsCopy := make([]v1.NodeCondition, len(conditions))
+	copy(conditionsCopy, conditions)
```

Comment:
Aren't we generating conditions in the same deterministic order?
If so, maybe we can avoid sorting them?

@yujuhong - wdyt?

Reply:
Kubelet does generate node conditions in a deterministic order, but there are other components that could patch the node status (e.g., NPD).

To be safe, I think we should still sort the conditions.